standard/arrays.md

```diff
@@ -46,9 +46,18 @@ In effect, the *rank_specifier*s are read from left to right *before* the final
 
 > *Example*: The type in `T[][,,][,]` is a single-dimensional array of three-dimensional arrays of two-dimensional arrays of `int`. *end example*
 
-At run-time, a value of an array type can be `null` or a reference to an instance of that array type.
+An array type is a reference type and therefore can be a nullable reference type ([§8.9.3](types.md#893-nullable-reference-types)). Likewise the element type of an array can be a nullable reference type or a nullable value type ([§8.3.12](types.md#8312-nullable-value-types)):
 
-> *Note*: Following the rules of [§17.6](arrays.md#176-array-covariance), the value may also be a reference to a covariant array type. *end note*
+- An array type of the form `T[R]` is a non-nullable array with rank `R` and a non-array non-nullable element type `T`.
+- An array type of the form `T[R]?` is a nullable array with rank `R` and a non-array non-nullable element type `T`.
+- An array type of the form `T?[R]` is a non-nullable array with rank `R` and a non-array nullable element type `T`.
+- An array type of the form `T?[R]?` is a nullable array with rank `R` and a non-array nullable element type `T`.
+
+At run-time, a value of an array type can be:
+
+- `null`; or
```

Comment:
I wonder whether it would be useful to have a note either here or in array creation expressions to warn that code that looks safe still can easily be null-unsafe with arrays:

```csharp
string[] x = new string[5];
int y = x[0].Length;
```

... no warnings, but will fail at run-time.

Reply:
@jskeet – Well there were no warnings from the current version of whatever compiler you used, but there is not reason why some compiler couldn’t issue a warning… I suggest you spin this off as a seperate issue.